File: CCVTAC/CCVTAC.Console/ExternalTools/Runner.cs
using System.Diagnostics;

namespace CCVTAC.Console.ExternalTools;

internal static class Runner
{
    internal static Result Run(ToolSettings settings, Printer printer)
    {
        Watch watch = new();

        printer.Print($"Starting {settings.Program.Name} for {settings.Program.Purpose}...");
        printer.Print($"Running command: {settings.Program.Name} {settings.Args}");

        ProcessStartInfo processStartInfo = new()
        {
            FileName = settings.Program.Name,
            Arguments = settings.Args,
            UseShellExecute = false,
            RedirectStandardOutput = false,
            RedirectStandardError = false,
            CreateNoWindow = true,
            WorkingDirectory = settings.WorkingDirectory
        };

        using Process? process = Process.Start(processStartInfo);

        if (process is null)
        {
            return Result.Fail($"Could not start {settings.Program.Name}. " +
                               $"Please install it from {settings.Program.Url}.");
        }

        process.WaitForExit();

        printer.Print($"Done with {settings.Program.Purpose} in {watch.ElapsedFriendly}.");

        int exitCode = process.ExitCode;
        if (exitCode == 0)
        {
            return Result.Ok();
        }

        if (settings.ExitCodes is not null &&
            settings.ExitCodes.ContainsKey(exitCode))
        {
            return Result.Fail($"External program \"{settings.Program.Name}\" reported error {exitCode}: {settings.ExitCodes[exitCode]}.");
        }

        return Result.Fail($"External program \"{settings.Program.Name}\" reported an error ({exitCode}).");
    }
}

File: src/CCVTAC.Main/ExternalTools/Runner.fs
namespace CCVTAC.Main.ExternalTools

open CCVTAC.Main
open Startwatch.Library
open System
open System.Diagnostics

module Runner =

    type ToolResult = { ExitCode: int; Error: string option }

    [<Literal>]
    let private authenticSuccessExitCode = 0

    let private isSuccessExitCode (otherSuccessExitCodes: int list) (exitCode: int) =
        List.contains exitCode (authenticSuccessExitCode :: otherSuccessExitCodes)

    /// Calls an external application.
    /// <param name="settings">Tool settings for execution</param>
    /// <param name="otherSuccessExitCodes">Additional exit codes, other than 0, that can be treated as non-failures</param>
    /// <param name="printer">Printer for logging</param>
    /// <returns>A Result instance containing the exit code and any warnings or else an error message.</returns>
    let runTool toolSettings otherSuccessExitCodes (printer: Printer)
        : Result<ToolResult, string> =

        let watch = Watch()
        printer.Info $"Running {toolSettings.CommandWithArgs}..."

        let command =
            toolSettings.CommandWithArgs.Split([|' '|], 2)
            |> fun arr -> {| Tool = arr[0]
                             Args = if Array.hasMultiple arr then arr[1] else String.Empty |}

        let processStartInfo = ProcessStartInfo command.Tool
        processStartInfo.Arguments <- command.Args
        processStartInfo.RedirectStandardOutput <- false
        processStartInfo.RedirectStandardError <- true
        processStartInfo.CreateNoWindow <- true
        processStartInfo.WorkingDirectory <- toolSettings.WorkingDirectory

        match Process.Start processStartInfo with
        | Null ->
            Error $"Could not locate or start {command.Tool}."
        | NonNull process' ->
            let error = process'.StandardError.ReadToEnd()

            process'.WaitForExit()
            printer.Info $"{command.Tool} finished in {watch.ElapsedFriendly}."

            let trimmedError = if String.hasText error
                               then Some (String.trimTerminalLineBreak error)
                               else None

            if isSuccessExitCode otherSuccessExitCodes process'.ExitCode
            then Ok { ExitCode = process'.ExitCode; Error = trimmedError }
            else Error $"{command.Tool} exited with code {process'.ExitCode}: {trimmedError}."
